src/chunk_constraint.c

```diff
@@ -651,17 +651,23 @@ chunk_constraint_delete_metadata(TupleInfo *ti)
 		heap_getattr(ti->tuple, Anum_chunk_constraint_constraint_name, ti->desc, &isnull);
 	int32 chunk_id =
 		DatumGetInt32(heap_getattr(ti->tuple, Anum_chunk_constraint_chunk_id, ti->desc, &isnull));
-	Chunk *chunk = ts_chunk_get_by_id(chunk_id, 0, true);
-	Oid index_relid = get_constraint_index(
-		get_relation_constraint_oid(chunk->table_id, NameStr(*DatumGetName(constrname)), true));
+	/* Get the chunk relid. Note that, at this point, the chunk table can be
+	 * deleted already */
+	Oid chunk_relid = ts_chunk_get_relid(chunk_id, true);
```

Comment:
I don't understand the reason for the code change in this file : use ts_chunk_get_relid versus ts_chunk_get_by_id. Could you elaborate why the code is modified here?  It looks like ts_chunk_get_by_id will throw an error if we try to do this on a  chunk that is marked dropped. Isn't that the desired behavior?

Reply:
There are a couple of reasons. The main one is that I added some Asserts for hardening that is now hit without this function. For example, there's now and Assert that checks that a Chunk cannot have an invalid `table_id`, i.e., you cannot have a Chunk object without the table actually existing. When you DROP a hypertable and it cascades to chunks, the data table is actually removed before we remove the metadata, so this assert would be hit. Thus, we'd need a way to read chunk metadata rows without building complete chunks.

`ts_chunk_get_by_id` allocates a "complete" chunk, and joins in other metadata like constraints, relids for the data tables, etc. When dropping a chunk, this function is called and in various places over-and-over again when the drop cascades to other metadata, like `chunk_constraint_delete_metadata` and `chunk_constraint_drop_constraint`. In contrast, `ts_chunk_get_relid` only reads the minimal data

Ideally, we'd never be able to build `Chunk` objects when they don't really exists (e.g., we should never have a `Chunk` object that has `dropped=true`) but this requires a major refactoring of the code.
This is a lot of unnecessary work that's not needed in many places. In contrast,  `ts_chunk_get_relid` translates the internal ID to a relid by reading _only_ the chunk row metadata.